server/src/main/java/io/unitycatalog/server/service/credential/aws/AwsCredentialVendor.java

```diff
@@ -0,0 +1,81 @@
+import java.time.Duration;
+import java.util.Map;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.model.Credentials;
+
+public class AwsCredentialVendor {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AwsCredentialVendor.class);
+
+  private final Map<String, S3StorageConfig> s3Configurations;
+
+  public AwsCredentialVendor() {
+    this.s3Configurations = ServerPropertiesUtils.getInstance().getS3Configurations();
+  }
+
+  public Credentials vendAwsCredentials(CredentialContext context) {
+    S3StorageConfig s3StorageConfig = s3Configurations.get(context.getStorageBase());
+    if (s3StorageConfig == null) {
+      throw new BaseException(ErrorCode.FAILED_PRECONDITION, "S3 bucket configuration not found.");
+    }
+
+    if (s3StorageConfig.getSessionToken() != null && !s3StorageConfig.getSessionToken().isEmpty()) {
+      // if a session token was supplied, then we will just return static session credentials
+      return Credentials.builder()
+          .accessKeyId(s3StorageConfig.getAccessKey())
+          .secretAccessKey(s3StorageConfig.getSecretKey())
+          .sessionToken(s3StorageConfig.getSessionToken())
+          .build();
+    }
+
+    // TODO: cache sts client
+    StsClient stsClient = getStsClientForStorageConfig(s3StorageConfig);
+
+    // TODO: Update this with relevant user/role type info once available
+    String roleSessionName = "uc-%s".formatted(UUID.randomUUID());
+    String awsPolicy =
+        AwsPolicyGenerator.generatePolicy(context.getPrivileges(), context.getLocations());
+
+    return stsClient
+        .assumeRole(
+            r ->
+                r.roleArn(s3StorageConfig.getAwsRoleArn())
+                    .policy(awsPolicy)
+                    .roleSessionName(roleSessionName)
+                    .durationSeconds((int) Duration.ofHours(1).toSeconds()))
+        .credentials();
+  }
+
+  private StsClient getStsClientForStorageConfig(S3StorageConfig s3StorageConfig) {
+    AwsCredentialsProvider credentialsProvider;
+    if (s3StorageConfig.getSecretKey() != null && !s3StorageConfig.getAccessKey().isEmpty()) {
+      credentialsProvider =
+          StaticCredentialsProvider.create(
+              AwsBasicCredentials.create(
+                  s3StorageConfig.getAccessKey(), s3StorageConfig.getSecretKey()));
+    } else {
+      credentialsProvider = DefaultCredentialsProvider.create();
+    }
+
+    // TODO: should we try and set the region to something configurable or specific to the server
+    // instead?
```

Comment:
@alexreid-db Do you mean setting the region to be based on the UC server instance instead of making it per storage config?

Reply:
yeah, just for the sts client